client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClientProxyHandler.kt

```
@@ -528,36 +530,36 @@ class RPCClientProxyHandler(
     }
 
     private fun attemptReconnect() {
-        var reconnectAttempts = rpcConfiguration.maxReconnectAttempts.times(serverLocator.staticTransportConfigurations.size)
+        // This can be a negative number as `rpcConfiguration.maxReconnectAttempts = -1` means infinite number of re-connects
+        val maxReconnectCount = rpcConfiguration.maxReconnectAttempts.times(serverLocator.staticTransportConfigurations.size)
+        log.debug { "maxReconnectCount = $maxReconnectCount" }
+        var reconnectAttempt = 1
         var retryInterval = rpcConfiguration.connectionRetryInterval
         val maxRetryInterval = rpcConfiguration.connectionMaxRetryInterval
 
-        var transportIterator = serverLocator.staticTransportConfigurations.iterator()
-        while (transportIterator.hasNext() && reconnectAttempts != 0) {
-            val transport = transportIterator.next()
-            if (!transportIterator.hasNext())
-                transportIterator = serverLocator.staticTransportConfigurations.iterator()
+        while (if (maxReconnectCount > 0) reconnectAttempt < maxReconnectCount else true) {
```

having an if conditional inside a while statement makes it a bit hard to parse mentally.
It'd be better to have a private helper method, smth like `while(shouldRetry(reconnectAttempt, maxReconnectCount))`

Also, shouldn't the logic be instead:
`if (maxReconnectCount >= 0) reconnectAttempt <= maxReconnectCount else true`?

---

Thanks @dimosr 
I guess it is down to the contract for `maxReconnectCount`, which is not written anywhere. From the code when `maxReconnectCount` is `-1` then this means infinite number of reconnects. But what about when `maxReconnectCount == 0` ?
Since it is not written, my assumption was that  `maxReconnectCount = 0` also means infinite number of re-connects.

I take your point about helper function and will make a change accordingly.

---

yeah, I also had a look at the docs, it's not completely clear, indeed. I assume that any non-negative values would have the natural meaning, so 0 would mean no retries.

I wouldn't be against the current logic assuming we'll clarify the docs to make clear the special meaning of 0. Another question I'd have in this case is is there any way to remove retries if we take the option of setting it to zero away?

---

Sure - I will update the docs to be explicit.

> is there any way to remove retries
I guess there is a little bit of contradiction here. If the user providing a pool of endpoints of size greater than 1, why would such a user ever ask for no retries to happen?

---

that's a good point, not sure. Maybe if someone wants to have all addresses tried on the first connection attempt, while at the same time failing fast and avoiding any retries on transient issues?

---

Yes, I agree - but this seem to be a very unusual type of requirement.

---

And I have updated the KDocs on `maxReconnectAttempts` to explicty say: `Any non-positive value would mean that there will be an infinite number of reconnect attempts.`

---

Given this is essentially a refactoring change, I'd prefer to keep the old behaviour since there might be customers relying on it already.

---

I still think we have to explicitly state what is going to happen when a negative or zero value is passed into `maxReconnectAttempts`. Especially given that in our documentation we say: 
"See `CordaRPCClientConfiguration.maxReconnectAttempts` for adjusting the number of retries."

I doubt any of customers put a value of `0` to `maxReconnectAttempts`, but @dimosr do you think it would be correct to change the Kdocs to say `Any negative value would mean that there will be an infinite number of reconnect attempts.` and change the code accordingly?

---

In cases like this, I usually prefer to be on the safe side and preserve existing behaviour (just in case someone relies on it) unless of course there's a clear requirement to change it. So, I'd do the following here:
* Adjust the kdocs to make clear what's the contract for negative values, i.e. infinite retries as you suggested. I believe that's already what the code does, so no code changes needed there.
* Change the code, so that it does not perform retries when `maxReconnectAttempts = 0`, similar to the existing behaviour. On the kdocs side, either call this out - the fact that zero means no retries - or not mention anything at all. I think this is the natural meaning of this value, so it shouldn't be a big problem.

---

OK, I will update the KDoc and change the code such that we do not re-connect if `maxReconnectAttempts = 0`.

---

@dimosr - I have pushed more changes as https://github.com/corda/corda/pull/6426/commits/cee7e796853b62f49abec0e681e8ce9f19488d24
Could you please re-review when you have time?